libs/ml/include/ml/layers/fully_connected.hpp

```diff
@@ -189,14 +197,44 @@ class FullyConnected : public SubGraph<T>
             inputs.front()->shape(inputs.front()->shape().size() - 1)};
   }
 
+  std::vector<SizeType> DefaultOutputShape() override
+  {
+    std::vector<SizeType> a{{out_size_}, {1}};
+    SubGraph<T>::default_output_shape_ = a;
+    return SubGraph<T>::default_output_shape_;
+  }
+
   static constexpr OpType OpCode()
   {
     return fetch::ml::OpType::LAYER_FULLY_CONNECTED;
   }
 
+  typename fetch::ml::ops::Ops<T>::ChargeAmount OpForwardCost(
+      typename fetch::ml::ops::Ops<T>::VecShapesType const &input_shapes) override
+  {
+    FETCH_UNUSED(input_shapes);
+    // Input shape is ignored because Dense layer knows all shapes already,
+    // however it's a good idea to check if given shape matches to prevent crash.
+
+    FETCH_LOG_INFO(DESCRIPTOR,
+                   "Calculating forward pass cost: " + ops::Ops<T>::PrintMyOutputShape());
+
+    // TODO(VH): stop on Placeholder, and calculate correct input shape for each
+    // layer (assume Flatten, MatMul, Add, Activation are worth calculating)
+    typename fetch::ml::ops::Ops<T>::ChargeAmount total_cost = 0;
+    for (std::pair<std::string, typename Graph<T>::NodePtrType> const &node : nodes_)
+    {
+      total_cost += node.second->GetOp()->OpForwardCost({{out_size_, in_size_}});
+    }
+
+    FETCH_LOG_INFO(DESCRIPTOR, "  Cost calculated : " + std::to_string(total_cost) + "\n");
+    return total_cost;
+  }
+
   static constexpr char const *DESCRIPTOR = "FullyConnected";
 
 private:
+  using Graph<T>::nodes_;
```

Comment:
please don't do this, just use this->nodes_ because it makes the intention clearer

Reply:
removed